--- sendmail-8.12.0/devtools/OS/Linux.orig	Mon Sep 24 09:48:15 2001
+++ sendmail-8.12.0/devtools/OS/Linux	Mon Sep 24 09:56:53 2001
@@ -1,9 +1,10 @@
 #	$Id$
 define(`confDEPEND_TYPE', `CC-M')
 define(`confSM_OS_HEADER', `sm_os_linux')
-define(`confMANROOT', `/usr/man/man')
+define(`confMANROOT', `/usr/share/man/man')
 define(`confLIBS', `-ldl')
 define(`confEBINDIR', `/usr/sbin')
+define(`confSTDIR', `/var/log')
 APPENDDEF(`confLIBSEARCH', `crypt nsl')
 
 define(`confLD', `ld')
