Index: proto.m4
===================================================================
RCS file: /cvs/cf/m4/proto.m4,v
retrieving revision 8.649.2.12
retrieving revision 8.649.2.13
diff -u -r8.649.2.12 -r8.649.2.13
--- cf/m4/proto.m4	3 Dec 2002 16:48:37 -0000	8.649.2.12
+++ cf/m4/proto.m4	4 Dec 2002 00:12:18 -0000	8.649.2.13
@@ -1761,6 +1761,9 @@
 dnl workspace: {client_name} $| {client_addr}
 R$+ $| $+		$: $>D < $1 > <?> <+ Connect> < $2 >
 dnl workspace: <result-of-lookup> <{client_addr}>
+dnl OR $| $+ if client_name is empty
+R   $| $+		$: $>A < $1 > <?> <+ Connect> <>	empty client_name
+dnl workspace: <result-of-lookup> <{client_addr}>
 R<?> <$+>		$: $>A < $1 > <?> <+ Connect> <>	no: another lookup
 dnl workspace: <result-of-lookup> (<>|<{client_addr}>)
 R<?> <$*>		$: OK				found nothing
