diff -urNp -x '*.orig' sendmail-8.16.1.org/rmail/rmail.c sendmail-8.16.1/rmail/rmail.c
--- sendmail-8.16.1.org/rmail/rmail.c	2020-05-19 21:54:33.000000000 +0200
+++ sendmail-8.16.1/rmail/rmail.c	2021-10-03 22:30:57.328925167 +0200
@@ -276,7 +276,6 @@ main(argc, argv)
 	args[i++] = _PATH_SENDMAIL;	/* Build sendmail's argument list. */
 	args[i++] = "-G";		/* relay submission */
 	args[i++] = "-oee";		/* No errors, just status. */
-	args[i++] = "-odq";		/* Queue it, don't try to deliver. */
 	args[i++] = "-oi";		/* Ignore '.' on a line by itself. */
 
 	/* set from system and protocol used */
