diff -urNp -x '*.orig' sendmail-8.16.1.org/devtools/OS/Linux sendmail-8.16.1/devtools/OS/Linux
--- sendmail-8.16.1.org/devtools/OS/Linux	2020-05-19 21:54:32.000000000 +0200
+++ sendmail-8.16.1/devtools/OS/Linux	2021-10-03 22:30:57.502258314 +0200
@@ -6,9 +6,10 @@ dnl	Place personal settings in devtools/
 define(`confDEPEND_TYPE', `CC-M')
 define(`confCCOPTS_SO', `-fPIC')
 define(`confSM_OS_HEADER', `sm_os_linux')
-define(`confMANROOT', `/usr/man/man')
+define(`confMANROOT', `/usr/share/man/man')
 define(`confLIBS', `-ldl')
 define(`confEBINDIR', `/usr/sbin')
+define(`confSTDIR', `/var/log')
 APPENDDEF(`confLIBSEARCH', `crypt nsl')
 
 define(`confLD', `ld')
