diff -Nur sendmail-8.9.1/src/pathnames.h sendmail-8.9.1.orig/src/pathnames.h
--- sendmail-8.9.1/src/pathnames.h	Wed May 20 01:36:02 1998
+++ sendmail-8.9.1.orig/src/pathnames.h	Sat Sep 12 14:49:03 1998
@@ -15,7 +15,7 @@
 # if defined(USE_VENDOR_CF_PATH) && defined(_PATH_VENDOR_CF)
 #  define _PATH_SENDMAILCF	_PATH_VENDOR_CF
 # else
-#  define _PATH_SENDMAILCF	"/etc/sendmail.cf"
+#  define _PATH_SENDMAILCF	"/etc/mail/sendmail.cf"
 # endif
 #endif
 
