diff -Nuard sendmail-8.12.9.orig/sendmail/parseaddr.c sendmail-8.12.9/sendmail/parseaddr.c
--- sendmail-8.12.9.orig/sendmail/parseaddr.c	2003-03-27 03:39:53.000000000 +0100
+++ sendmail-8.12.9/sendmail/parseaddr.c	2003-09-17 00:51:40.000000000 +0200
@@ -700,7 +700,12 @@
 						addr[MAXNAME] = '\0';
 	returnnull:
 					if (delimptr != NULL)
+					{
+						if (p > addr)
+							p++;
+							
 						*delimptr = p;
+					}
 					CurEnv->e_to = saveto;
 					return NULL;
 				}
