diff -Nur sendmail-8.12.0.orig/cf/cf/Makefile sendmail-8.12.0/cf/cf/Makefile
--- sendmail-8.12.0.orig/cf/cf/Makefile	Mon Aug 20 17:16:48 2001
+++ sendmail-8.12.0/cf/cf/Makefile	Wed Sep 26 01:33:48 2001
@@ -76,10 +76,10 @@
 install-cf:  install-sendmail-cf install-submit-cf
 
 install-sendmail-cf: $(CF).cf
-	$(INSTALL) -c -o $(CFOWN) -g $(CFGRP) -m $(CFMODE) $(CF).cf ${DESTDIR}$(MAILDIR)/sendmail.cf
+	$(INSTALL) -c -m $(CFMODE) $(CF).cf ${DESTDIR}$(MAILDIR)/sendmail.cf
 
 install-submit-cf: $(SUBMIT).cf
-	$(INSTALL) -c -o $(CFOWN) -g $(CFGRP) -m $(CFMODE) $(SUBMIT).cf ${DESTDIR}$(MAILDIR)/submit.cf
+	$(INSTALL) -c -m $(CFMODE) $(SUBMIT).cf ${DESTDIR}$(MAILDIR)/submit.cf
 
 depend:
 
diff -Nur sendmail-8.12.0.orig/sendmail/Makefile.m4 sendmail-8.12.0/sendmail/Makefile.m4
--- sendmail-8.12.0.orig/sendmail/Makefile.m4	Sat Jun 16 18:54:49 2001
+++ sendmail-8.12.0/sendmail/Makefile.m4	Wed Sep 26 01:35:09 2001
@@ -11,7 +11,7 @@
 
 dnl hack: /etc/mail is not defined as "location of .cf" in the build system
 define(`bldTARGET_INST_DEP', ifdef(`confINST_DEP', `confINST_DEP',
-`/etc/mail/submit.cf ${MSPQ}'))dnl
+`${DESTDIR}/etc/mail/submit.cf ${DESTDIR}${MSPQ}'))dnl
 define(`bldTARGET_LINKS', ifdef(`confLINKS', `confLINKS',
 `${DESTDIR}${UBINDIR}/newaliases ${DESTDIR}${UBINDIR}/mailq ${DESTDIR}${UBINDIR}/hoststat ${DESTDIR}${UBINDIR}/purgestat')
 )dnl
@@ -36,14 +36,14 @@
 statistics:
 	${CP} /dev/null statistics
 
-/etc/mail/submit.cf:
+${DESTDIR}/etc/mail/submit.cf:
 	@echo "Please read INSTALL if anything fails while installing the binary."
 	@echo "/etc/mail/submit.cf will be installed now."
 	cd ${SRCDIR}/cf/cf && make install-submit-cf
 
 MSPQ=ifdef(`confMSP_QUEUE_DIR', `confMSP_QUEUE_DIR', `/var/spool/clientmqueue')
 
-${MSPQ}:
+${DESTDIR}${MSPQ}:
 	@echo "Please read INSTALL if anything fails while installing the binary."
 	@echo "You must have setup a new user ${MSPQOWN} and a new group ${GBINGRP}"
 	@echo "as explained in sendmail/SECURITY."
@@ -61,7 +61,7 @@
 divert(bldTARGETS_SECTION)
 
 install-set-user-id: bldCURRENT_PRODUCT ifdef(`confNO_HELPFILE_INSTALL',, `install-hf') ifdef(`confNO_STATISTICS_INSTALL',, `install-st') ifdef(`confNO_MAN_BUILD',, `install-docs')
-	${INSTALL} -c -o ${S`'BINOWN} -g ${S`'BINGRP} -m ${S`'BINMODE} bldCURRENT_PRODUCT ${DESTDIR}${M`'BINDIR}
+	${INSTALL} -c -m ${S`'BINMODE} bldCURRENT_PRODUCT ${DESTDIR}${M`'BINDIR}
 	for i in ${sendmailTARGET_LINKS}; do \
 		rm -f $$i; \
 		ln -s ${M`'BINDIR}/sendmail $$i; \
@@ -69,7 +69,7 @@
 
 define(`confMTA_LINKS', `${DESTDIR}${UBINDIR}/newaliases ${DESTDIR}${UBINDIR}/mailq ${DESTDIR}${UBINDIR}/hoststat ${DESTDIR}${UBINDIR}/purgestat')
 install-sm-mta: bldCURRENT_PRODUCT
-	${INSTALL} -c -o ${M`'BINOWN} -g ${M`'BINGRP} -m ${M`'BINMODE} bldCURRENT_PRODUCT ${DESTDIR}${M`'BINDIR}/sm-mta
+	${INSTALL} -c -m ${M`'BINMODE} bldCURRENT_PRODUCT ${DESTDIR}${M`'BINDIR}/sm-mta
 	for i in confMTA_LINKS; do \
 		rm -f $$i; \
 		ln -s ${M`'BINDIR}/sm-mta $$i; \
